### PR TITLE
Add configuration option to disable system metrics collection

### LINKED ISSUE #2052: Add a flag to disable system metrics collection
### 🚀 The feature

Add a flag to disable system metrics collection.

Or add a config file to control system metrics collection.

### Motivation, pitch

We notice that there is no option to control system metrics collection (if not, please let me know).

**Redundancy**

In our case, system metrics collection is redundant because we are already collecting system metrics using other tools like DCGM exporter

**Metrics collection is not working on MIG GPUS** 

Also, we are using MIG GPUs, and MIG GPUs have different UUID and accessibility (maybe https://github.com/pytorch/serve/issues/1237 is similar issue). So metrics collection is not working on MIG GPUs.

```sh
File "ts/metrics/metric_collector.py", line 27, in <module> 
     system_metrics.collect_all(sys.modules['ts.metrics.system_metrics'], arguments.gpu) 
   File "/home/venv/lib/python3.8/site-packages/ts/metrics/system_metrics.py", line 119, in collect_all 
     value(num_of_gpu) 
   File "/home/venv/lib/python3.8/site-packages/ts/metrics/system_metrics.py", line 71, in gpu_utilization 
     info = nvgpu.gpu_info() 
   File "/home/venv/lib/python3.8/site-packages/nvgpu/__init__.py", line 8, in gpu_info 
     gpu_infos = [re.match('GPU ([0-9]+): ([^(]+) \(UUID: ([^)]+)\)', gpu).groups() for gpu in gpus] 
   File "/home/venv/lib/python3.8/site-packages/nvgpu/__init__.py", line 8, in <listcomp> 
     gpu_infos = [re.match('GPU ([0-9]+): ([^(]+) \(UUID: ([^)]+)\)', gpu).groups() for gpu in gpus] 
 AttributeError: 'NoneType' object has no attribute 'groups'
```

**Other issues we've experienced**  

We experienced that if multiple torchserve containers (~30) are starting simultaneously in a single GPU machine, `nvidia-smi` command hangs. We analyzed the issue and `nvidia-smi` call in metrics collector seems related.

Torchserve uses `nvgpu` to collect gpu infos and `nvgpu` calls `nvidia-smi`.
- https://github.com/rossumai/nvgpu/blob/v0.9.0/nvgpu/__init__.py#L6-L27

So, starting multiple containers can cause a lot of `nvidia-smi` calls and eventually `nvidia-smi` command hangs.

Please consider the request, thank you

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by namannandan**
## Description

Support disabling system metrics collection using the configuration option `disable_system_metrics` via a configuration file or the environment variable `TS_DISABLE_SYSTEM_METRICS`.

Fixes #2052

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Feature/Issue validation/testing

- [ ] Regression tests
```
test_metrics.py::test_collect_system_metrics_when_not_disabled PASSED                                                                                                                                       [ 88%]
test_metrics.py::test_disable_system_metrics_using_config_properties PASSED                                                                                                                                 [ 94%]
test_metrics.py::test_disable_system_metrics_using_environment_variable PASSED                                                                                                                              [100%]
```
[cpu_regression_system_metrics_disable.log](https://github.com/pytorch/serve/files/10680532/cpu_regression_system_metrics_disable.log)
[gpu_regression_system_metrics_disable.log](https://github.com/pytorch/serve/files/10680533/gpu_regression_system_metrics_disable.log)

- [ ] Manual test with system metrics disabled. Register a model and make inference requests
```
$ cat logs/ts_metrics.log 
2023-02-02T00:56:55,064 - Requests2XX.Count:1|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328214
2023-02-02T00:57:25,592 - Requests2XX.Count:1|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328214
2023-02-02T00:57:29,198 - W-9001-resnet18_1.0.ms:3608|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328249
2023-02-02T00:57:29,198 - W-9000-resnet18_1.0.ms:3609|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328249
2023-02-02T00:57:29,199 - WorkerThreadTime.ms:24|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328249
2023-02-02T00:57:29,199 - WorkerThreadTime.ms:24|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328249
2023-02-02T00:58:45,775 - Requests2XX.Count:1|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328214
2023-02-02T00:58:45,777 - QueueTime.ms:0|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328325
2023-02-02T00:58:45,778 - WorkerThreadTime.ms:7|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328325
2023-02-02T01:02:45,042 - Requests2XX.Count:1|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328214
2023-02-02T01:02:45,042 - QueueTime.ms:0|#Level:Host|#hostname:88665a372f4b.ant.amazon.com,timestamp:1675328565
```
System metrics such as `CPUUtilization`, `MemoryUtilization` etc.. are not collected and reported

- [ ] Manual test to verify config snapshot
```
$ cat custom_config.properties 
disable_system_metrics=true
$ ls ../model_store 
resnet-18-py-eager.mar
$  rm -rf logs
$ torchserve --start --model-store ../model_store --models all --ts-config custom_config.properties
.....
.....
Number of GPUs: 0
Number of CPUs: 12
Max heap size: 4096 M
Python executable: /Users/namannan/.pyenv/versions/3.8.13/bin/python
Config file: custom_config.properties
Inference address: http://127.0.0.1:8080
Management address: http://127.0.0.1:8081
Metrics address: http://127.0.0.1:8082
Model Store: /Volumes/workplace/pytorch/model_store
Initial Models: all
Log dir: /Volumes/workplace/pytorch/serve/logs
Metrics dir: /Volumes/workplace/pytorch/serve/logs
Netty threads: 0
Netty client threads: 0
Default workers per model: 12
Blacklist Regex: N/A
Maximum Response Size: 6553500
Maximum Request Size: 6553500
Limit Maximum Image Pixels: true
Prefer direct buffer: false
Allowed Urls: [file://.*|http(s)?://.*]
Custom python dependency for model allowed: false
Metrics report format: prometheus
Enable metrics API: true
Disable system metrics: true
Workflow Store: /Volumes/workplace/pytorch/model_store
Model config: N/A
.....
.....
$ curl -X GET "http://localhost:8081/models"
{
  "models": [
    {
      "modelName": "resnet-18-py-eager",
      "modelUrl": "resnet-18-py-eager.mar"
    }
  ]
}
$ torchserve --stop
$ torchserve --start --model-store ../model_store --models all
.....
.....
Number of GPUs: 0
Number of CPUs: 12
Max heap size: 4096 M
Python executable: /Users/namannan/.pyenv/versions/3.8.13/bin/python
Config file: logs/config/20230207163503771-shutdown.cfg
Inference address: http://127.0.0.1:8080
Management address: http://127.0.0.1:8081
Metrics address: http://127.0.0.1:8082
Model Store: /Volumes/workplace/pytorch/model_store
Initial Models: all
Log dir: /Volumes/workplace/pytorch/serve/logs
Metrics dir: /Volumes/workplace/pytorch/serve/logs
Netty threads: 0
Netty client threads: 0
Default workers per model: 12
Blacklist Regex: N/A
Maximum Response Size: 6553500
Maximum Request Size: 6553500
Limit Maximum Image Pixels: true
Prefer direct buffer: false
Allowed Urls: [file://.*|http(s)?://.*]
Custom python dependency for model allowed: false
Metrics report format: prometheus
Enable metrics API: true
Disable system metrics: true
Workflow Store: /Volumes/workplace/pytorch/model_store
Model config: N/A
2023-02-07T16:35:43,958 [INFO ] main org.pytorch.serve.snapshot.SnapshotManager - Started restoring models from snapshot {
  "name": "20230207163503771-shutdown.cfg",
  "modelCount": 1,
  "created": 1675816503771,
  "models": {
    "resnet-18-py-eager": {
      "1.0": {
        "defaultVersion": true,
        "marName": "resnet-18-py-eager.mar",
        "minWorkers": 12,
        "maxWorkers": 12,
        "batchSize": 1,
        "maxBatchDelay": 100,
        "responseTimeout": 120
      }
    }
  }
}
.....
.....
$ curl -X GET "http://localhost:8081/models"
{
  "models": [
    {
      "modelName": "resnet-18-py-eager",
      "modelUrl": "resnet-18-py-eager.mar"
    }
  ]
}
$ torchserve --stop
$ torchserve --start --model-store ../model_store --models all
$ curl -X GET "http://localhost:8081/models"
{
  "models": [
    {
      "modelName": "resnet-18-py-eager",
      "modelUrl": "resnet-18-py-eager.mar"
    }
  ]
}
$ torchserve --stop
```

## Checklist:

- [x] Did you have fun?
- [x] Have you added tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: test/pytest/test_metrics.py [hunk 1]
```diff
@@ -318,3 +318,70 @@ def test_metrics_location_var_snapshot_enabled_rdonly_dir():
...
+        os.remove(f)
+
+    config_file = test_utils.ROOT_DIR + "config.properties"
+    with open(config_file, "w+") as f:
+        f.write("disable_system_metrics=true")
+
+    try:
+        test_utils.start_torchserve(snapshot_file=config_file)
+        assert len(glob.glob("logs/ts_metrics.log")) == 1
+        assert os.path.getsize(glob.glob("logs/ts_metrics.log")[0]) == 0
```
Comment (msaroufim) [inline test/pytest/test_metrics.py:359]: So the test right now seems to check that logs still exist even if the system metrics don't, a better test might be to also explicitly check that some strings like `CPUUtilization` don't show up in the logs anymore

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: test/pytest/test_metrics.py (hunk 1 above)
Comment (namannandan) [reply test/pytest/test_metrics.py:359]: Sounds good, will update the test to specifically check for system metrics

Comment (namannandan) [reply test/pytest/test_metrics.py]: Updated tests to check for system metric names

**Review by msaroufim — APPROVED** [review]

File: docs/configuration.md [hunk 2]
```diff
@@ -228,7 +228,7 @@ Set nvidia environment variables. For example:
     }
 }
 ```
-A model's parameters are defined in [model source code](https://github.com/pytorch/serve/blob/master/frontend/server/src/main/java/org/pytorch/serve/wlm/Model.java#L24)
+A model's parameters are defined in [model source code](https://github.com/pytorch/serve/blob/05e80bebbca8e5c152e0b66977f8c7775a44b803/frontend/server/src/main/java/org/pytorch/serve/wlm/Model.java#L24)
```
Comment (lxning) [inline docs/configuration.md:231]: don't change the link

**Review by lxning — COMMENTED** [review]
- please attach complete regression test log
- check if snapshot works fine. (ie. run "torchserve --start --model-store xxx --models all" twice)

File: docs/configuration.md (hunk 2 above)
Comment (namannandan) [reply docs/configuration.md]: Reverted the link

**Comment by namannandan** [comment]
Added CPU and GPU regression logs in the summary. Also included the transcript for a manual test session to check that config snapshot works as expected.

**Review by agunapal — APPROVED** [review]

**Pull request merged**